File: Data Structures & Algorithms/reverse-a-linked-list/submission-0.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* reverseList(ListNode* head) {
        if (head == nullptr) return nullptr;

        stack<ListNode*> nodeStack; 
        ListNode* current = head;

        while(current) {
            nodeStack.push(current);
            current = current->next;
        }

        ListNode* newHead = nodeStack.top();
        nodeStack.pop();
        current = newHead;

        while(!nodeStack.empty()) {
            current->next = nodeStack.top();
            nodeStack.pop(); 

            current = current->next;
        }

        current->next = nullptr;
        
        return newHead; 
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-0.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

public class Solution {
    public void ReorderList(ListNode head) {
        if( head == null) return;

        var nodeStack = new Stack<ListNode>();
        var cur = head;

        while(cur != null) {
            nodeStack.Push(cur);
            cur = cur.next; 
        }

        int count = nodeStack.Count;
        var l = head; 

        for (int i = 0; i < count / 2; i++) {
            var r = nodeStack.Pop();
   
            var temp = l.next;

            l.next = r;
            r.next = temp;

            l = temp;
        }

        l.next = null;
    }
}
